### PR TITLE
feat(storage-browser): add custom actions

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes
Add custom actions support for `StorageBrowser`
- loosen handler interfaces to improve custom action ergonomics
- add `useHandler`, `useAction` and `useList` hooks
- simplify `useView`
- integrate  `useAction` and `useList` hooks in to `StorageBrowser` views
- refactor action config types

Usage:
```ts

import { ActionHandler, ActionViewConfig, createStorageBrowser } from '@aws-amplify/ui-react-storage';

type MyAction = ActionHandler<{ myOptionalData: string }, { myOptionalOutput: string }>
const myAction: MyAction =  () => {}

const myActionViewConfig: ActionViewConfig<MyAction, 'MyActionView'> = {}

const { useAction, StorageBrowser } = createStorageBrowser({ actions: { 
  // no view slot, just handler
  myAction,
  // with view slot
  myActionViewConfig
))

const MyActionView = () => {
  const [atomicState, atomicHandler] = useAction('myAction')

  const items = []
  const [batchState, batchHandler] = useAction('myAction', { items })
 
  return ( ... )
}

<StorageBrowser views={{ MyActionView }} />
```

#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] `yarn test` passes and tests are updated/added
- [ ] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/components/StorageBrowser/actions/handlers/delete.ts [hunk 1]
```diff
@@ -5,27 +5,34 @@ import {
...
 export interface DeleteHandler
   extends TaskHandler<DeleteHandlerInput, DeleteHandlerOutput> {}
 
 export const deleteHandler: DeleteHandler = ({
   config,
-  data: { key },
+  data,
 }): DeleteHandlerOutput => {
+  // eslint-disable-next-line no-console
+  console.log('delete data', data);
```
Comment (HuiSF) [inline packages/react-storage/src/components/StorageBrowser/actions/handlers/delete.ts:33]: Probably not intended?

File: packages/react-storage/src/components/StorageBrowser/useAction/useList.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  K extends keyof ListActionHooks,
+  S extends ListActionHooks[K],
+>(
+  type: K
+) => ReturnType<S>;
+
+const isListActionViewType = (value: unknown): value is ListActionType =>
+  Object.keys(LIST_ACTION_HOOKS).includes(value as ListActionType);
+
+// @ts-expect-error
```
Comment (HuiSF) [inline packages/react-storage/src/components/StorageBrowser/useAction/useList.ts:25]: sanity check: intended error suppressing?

File: packages/react-storage/src/components/StorageBrowser/actions/configs/defaults.tsx [hunk 3]
```diff
@@ -108,6 +77,5 @@ export const isDefaultActionViewType = (
 
 export const defaultActionConfigs = {
   ...defaultActionViewConfigs,
-  listLocationItems: listLocationItemsActionConfig,
-  listLocations: listLocationsActionConfig,
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/react-storage/src/components/StorageBrowser/actions/configs/defaults.tsx:112]: Apparently not?

File: packages/react-storage/src/components/StorageBrowser/actions/handlers/delete.ts (hunk 1 above)
Comment (cshfang) [reply packages/react-storage/src/components/StorageBrowser/actions/handlers/delete.ts:33]: Agreed

**Review by pranavosu — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]

File: packages/react-storage/src/components/StorageBrowser/StorageBrowserAmplify.tsx
Comment (AllanZhengYP) [inline packages/react-storage/src/components/StorageBrowser/StorageBrowserAmplify.tsx:1]: Non-blocking but I think this file is not used anywhere and can be cleared.

Comment (cshfang) [reply packages/react-storage/src/components/StorageBrowser/StorageBrowserAmplify.tsx:1]: I think you're probably right. We can do that separately for sure

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**